File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public static boolean check(int[] phash, int[] hash){
        for(int i = 0; i < phash.length; i++){
            if(phash[i] != hash[i]){
                return false;
            }
        }
        return true;
    }
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> ans = new ArrayList<>();
        if(s.length() < p.length()){
            return ans;            
        }
        int[] phash = new int[26];
        int[] hash = new int[26];
        int left = 0, right = 0, window = p.length();
        while(right < window){
            phash[p.charAt(right) - 'a']++;
            hash[s.charAt(right) - 'a']++;
            right++;
        }
        right -=1;
        while(right < s.length())
        {            
            boolean equal = check(phash, hash);
            if(equal)
            {
                ans.add(left);
            }
            right+=1;
            if(right != s.length())
            {
                hash[s.charAt(right) - 'a'] += 1;
            }
            hash[s.charAt(left) - 'a'] -=1 ;
            left += 1;
        }
        return ans;
    }
}

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> ans;
        if(s.length() < p.length()){
            return ans;            
        }
        vector<int> map1 (26, 0);
        vector<int> map2 (26, 0);
        int l = 0, r = 0;
        while(r < p.length()){
            map1[p[r] - 'a']++;
            map2[s[r] - 'a']++;
            r++;
        }
        
        if(map1 == map2){
            ans.push_back(l);
        }
        while(r < s.length()){
            map2[s[r] - 'a']++;
            map2[s[l] - 'a']--;
            l++; r++;
            if(map1 == map2){
                ans.push_back(l);
            }            
        }
        return ans;
    }
};